<?xml version="1.0"?>
<launch>

  <arg name="vehicle" default="iris"/>
  <arg name="ID" default="0"/>
  <arg name="localization" default="false" />
  <arg name="delete_data" default="true" />
  <arg name="rtabmap_viz" default="false"/>
  <arg name="database_path" default="~/.ros/rtabmap/px4/rtabmap.db" />

  <arg unless="$(arg delete_data)" name="args" default="" />
  <arg if="$(arg delete_data)" name="args" default="-d" />

  <arg name="scan_20_hz" default="false"/>
  <!-- If we launch the velodyne with "rpm:=1200" argument -->
  <arg name="deskewing" default="false"/>
  <arg name="slerp" default="false"/>
  <!-- If true, a slerp between the first and last time will be used to deskew each point, which is faster than using tf for every point but less accurate -->
  <arg name="scan_topic" default="/velodyne_points"/>
  <arg name="use_sim_time" default="false"/>
  <param if="$(arg use_sim_time)" name="use_sim_time" value="true"/>

  <arg name="imu_topic" default="/$(arg vehicle)_$(arg ID)/mavros/imu/data"/>
  <arg name="frame_id" default="base_link"/>
  <!-- base frame of the robot: for this example, we use velodyne as base frame -->
  <arg name="queue_size" default="100"/>
  <arg name="queue_size_odom" default="10"/>
  <arg name="loop_ratio" default="0.2"/>

  <arg name="resolution" default="0.1"/>
  <!-- set 0.05-0.3 for indoor, set 0.3-0.5 for outdoor -->
  <arg name="iterations" default="10"/>

  <!-- Grid parameters -->
  <arg name="ground_is_obstacle" default="true"/>
  <arg name="grid_max_range" default="10"/>

  <!-- For F2M Odometry -->
  <arg name="ground_normals_up" default="false"/>
  <!-- set to true when velodyne is always horizontal to ground (ground robot, car) -->
  <arg name="local_map_size" default="15000"/>
  <arg name="key_frame_thr" default="0.4"/>

  <!-- For FLOAM Odometry -->
  <arg name="floam" default="false"/>
  <!-- RTAB-Map should be built with FLOAM http://official-rtab-map-forum.206.s1.nabble.com/icp-odometry-with-LOAM-crash-tp8261p8563.html -->
  <arg name="floam_sensor" default="0"/>
  <!-- 0=16 rings (VLP16), 1=32 rings, 2=64 rings -->

  <!-- Lidar Deskewing -->
  <node if="$(arg deskewing)" pkg="nodelet" type="nodelet" name="lidar_deskewing" args="standalone rtabmap_util/lidar_deskewing" output="screen">
    <param name="wait_for_transform" value="0.1"/>
    <param name="fixed_frame_id" value="$(arg frame_id)"/>
    <param name="slerp" value="$(arg slerp)"/>
    <remap from="input_cloud" to="$(arg scan_topic)"/>
  </node>

  <arg if="$(arg deskewing)" name="scan_topic_deskewed" default="$(arg scan_topic)/deskewed"/>
  <arg unless="$(arg deskewing)" name="scan_topic_deskewed" default="$(arg scan_topic)"/>

  <group ns="rtabmap">
    <!-- <node pkg="rtabmap_odom" type="icp_odometry" name="icp_odometry" output="screen">
      <remap from="scan_cloud" to="$(arg scan_topic_deskewed)"/>
      <remap from="imu" to="$(arg imu_topic)"/>

      <param name="frame_id" type="string" value="$(arg frame_id)"/>
      <param name="odom_frame_id" type="string" value="odom"/>
      <param name="queue_size" type="int" value="$(arg queue_size_odom)"/>
      <param name="wait_imu_to_init" type="bool" value="true"/>
      <param name="wait_for_transform_duration" type="double" value="0.2"/>
      <param if="$(arg scan_20_hz)" name="expected_update_rate" type="double" value="25"/>
      <param unless="$(arg scan_20_hz)" name="expected_update_rate" type="double" value="20"/>
      <param name="publish_null_when_lost" type="bool" value="true"/>
      <param name="publish_tf" type="bool" value="false"/>


      <param name="Icp/PointToPlane" type="string" value="true"/>
      <param name="Icp/Iterations" type="string" value="$(arg iterations)"/>
      <param if="$(arg floam)" name="Icp/VoxelSize" type="string" value="0"/>
      <param unless="$(arg floam)" name="Icp/VoxelSize" type="string" value="$(arg resolution)"/>
      <param name="Icp/DownsamplingStep" type="string" value="1"/>

      <param name="Icp/Epsilon" type="string" value="0.001"/>
      <param if="$(arg floam)" name="Icp/PointToPlaneK" type="string" value="0"/>
      <param unless="$(arg floam)" name="Icp/PointToPlaneK" type="string" value="20"/>
      <param name="Icp/PointToPlaneRadius" type="string" value="0"/>
      <param name="Icp/MaxTranslation" type="string" value="2"/>
      <param name="Icp/MaxCorrespondenceDistance" type="string" value="$(eval resolution*10)"/>
      <param name="Icp/PM" type="string" value="true"/>
      <param name="Icp/PMOutlierRatio" type="string" value="0.7"/>
      <param name="Icp/CorrespondenceRatio" type="string" value="0.01"/>
      <param if="$(arg ground_normals_up)" name="Icp/PointToPlaneGroundNormalsUp" type="string" value="0.8"/>

      <param name="Odom/ResetCountdown" type="string" value="1"/>
      <param name="Odom/ScanKeyFrameThr" type="string" value="$(arg key_frame_thr)"/>
      <param if="$(arg floam)" name="Odom/Strategy" type="string" value="11"/>
      <param unless="$(arg floam)" name="Odom/Strategy" type="string" value="0"/>
      <param name="OdomF2M/ScanSubtractRadius" type="string" value="$(arg resolution)"/>
      <param name="OdomF2M/ScanMaxSize" type="string" value="$(arg local_map_size)"/>
      <param name="OdomLOAM/Sensor" type="string" value="$(arg floam_sensor)"/>
      <param name="OdomLOAM/Resolution" type="string" value="$(arg resolution)"/>
      <param if="$(eval not deskewing and scan_20_hz)" name="OdomLOAM/ScanPeriod" type="string" value="0.05"/>
      <param if="$(eval not deskewing and not scan_20_hz)" name="OdomLOAM/ScanPeriod" type="string" value="0.1"/>
      <param if="$(arg deskewing)" name="OdomLOAM/ScanPeriod" type="string" value="0"/>
    </node> -->

    <node pkg="rtabmap_slam" type="rtabmap" name="rtabmap" output="screen" args="$(arg args) --Vis/FeatureType 10 --Kp/DetectorStrategy 10 
            --Grid/MapFrameProjection true --NormalsSegmentation false --Grid/MaxGroundHeight 0 --Grid/MaxObstacleHeight 0
            --RGBD/StartAtOrigin true --Grid/VoxelSize 0.2 --Grid/DepthDecimation 1 --RGBD/OptimizeFromGraphEnd false --Rtabmap/StartNewMapOnLoopClosure true 
            --Optimizer/PriorsIgnored false --RGBD/LocalRadius 50"> 

      <param name="database_path" type="string" value="$(arg database_path)" />
      <param name="queue_size" type="int" value="$(arg queue_size)"/>
      <param name="publish_tf" type="bool" value="false"/>

      <param name="frame_id" type="string" value="$(arg frame_id)"/>
      <param name="subscribe_depth" type="bool" value="true"/>
      <param name="subscribe_rgb" type="bool" value="true"/>
      <param name="subscribe_scan_cloud" type="bool" value="true"/>
      <param name="approx_sync" type="bool" value="true"/>
      <param name="wait_for_transform_duration" type="double" value="0.2"/>

      <!-- <remap from="scan_cloud" to="assembled_cloud"/> -->
      <remap from="scan_cloud" to="$(arg scan_topic_deskewed)"/>
      <remap from="rgb/image" to="/camera/rgb/image_raw"/>
      <remap from="depth/image" to="/camera/depth/image_raw"/>
      <remap from="rgb/camera_info" to="/camera/rgb/camera_info"/>
      <remap from="imu" to="$(arg imu_topic)"/>
      <remap from="odom" to="/$(arg vehicle)_$(arg ID)/odom" />
      <remap from="gps/fix"  to="/$(arg vehicle)_$(arg ID)/mavros/global_position/global" />

      <remap from="grid_map" to="/map"/>

      <!-- RTAB-Map's parameters -->
      <param name="Rtabmap/DetectionRate" type="string" value="2"/>
      <param name="RGBD/NeighborLinkRefining" type="string" value="false"/>
      <param name="RGBD/ProximityBySpace" type="string" value="true"/>
      <param name="RGBD/ProximityMaxGraphDepth" type="string" value="0"/>
      <param name="RGBD/ProximityPathMaxNeighbors" type="string" value="1"/>
      <param name="RGBD/AngularUpdate" type="string" value="0.1"/>
      <param name="RGBD/LinearUpdate" type="string" value="0.1"/>
      <param name="Mem/NotLinkedNodesKept" type="string" value="false"/>
      <param name="Mem/STMSize" type="string" value="30"/>
      <param name="Mem/LaserScanNormalK" type="string" value="20"/>

      <param name="Reg/Strategy" type="string" value="1"/>
      <param name="Grid/FromDepth" type="bool" value="true"/>
      <param name="Grid/CellSize" type="string" value="$(arg resolution)"/>
      <param name="Grid/RangeMax" type="string" value="$(arg grid_max_range)"/>
      <param name="Grid/ClusterRadius" type="string" value="1"/>
      <param name="Grid/GroundIsObstacle" type="string" value="$(arg ground_is_obstacle)"/>
      <param name="Optimizer/GravitySigma" type="string" value="0.3"/>

      <!-- ICP parameters -->
      <param name="scan_voxel_size" type="string" value="$(arg resolution)"/>
      <param name="scan_normal_k" type="string" value="20"/>
      <param name="Icp/PointToPlaneRadius" type="string" value="0"/>
      <param name="Icp/PointToPlane" type="string" value="true"/>
      <param name="Icp/Iterations" type="string" value="$(arg iterations)"/>
      <param name="Icp/Epsilon" type="string" value="0.001"/>
      <param name="Icp/MaxTranslation" type="string" value="3"/>
      <param name="Icp/MaxCorrespondenceDistance" type="string" value="$(eval resolution*10)"/>
      <param name="Icp/Strategy" type="string" value="1"/>
      <param name="Icp/OutlierRatio" type="string" value="0.7"/>
      <param name="Icp/CorrespondenceRatio" type="string" value="$(arg loop_ratio)"/>

      <!-- localization mode -->
      <param if="$(arg localization)" name="Mem/IncrementalMemory" type="string" value="false" />
      <param unless="$(arg localization)" name="Mem/IncrementalMemory" type="string" value="true" />
      <param name="Mem/InitWMWithAllNodes" type="string" value="$(arg localization)" />
    </node>

    <node if="$(arg rtabmap_viz)" name="rtabmap_viz" pkg="rtabmap_viz" type="rtabmap_viz" output="screen">
      <param name="frame_id" type="string" value="$(arg frame_id)"/>
      <param name="odom_frame_id" type="string" value="odom"/>
      <param name="subscribe_odom_info" type="bool" value="true"/>
      <param name="subscribe_scan_cloud" type="bool" value="true"/>
      <param name="approx_sync" type="bool" value="false"/>
      <remap from="scan_cloud" to="odom_filtered_input_scan"/>
      <remap from="odom_info" to="odom_info"/>
    </node>

    <node pkg="nodelet" type="nodelet" name="point_cloud_aggregator" args="standalone rtabmap_util/point_cloud_aggregator" output="screen">
      <remap from="cloud1" to="/camera/depth/points"/>
      <remap from="cloud2" to="/velodyne_points"/>
      <remap from="combined_cloud" to="/rtabmap/fusion_cloud"/>
      <param name="xyz_output" type="bool" value="true" />
      <param name="approx_sync" type="bool" value="true" />
      <param name="frame_id" type="string" value="world" />
      <param name="fixed_frame_id" type="string" value="world" />
      <param name="queue_size" type="int" value="$(arg queue_size)" />
    </node>

    <!-- <node pkg="nodelet" type="nodelet" name="point_cloud_assembler" args="standalone rtabmap_util/point_cloud_assembler" output="screen">
      <remap from="cloud" to="$(arg scan_topic_deskewed)"/>
      <remap from="odom" to="odom"/>
      <param if="$(arg scan_20_hz)" name="max_clouds" type="int" value="20" />
      <param unless="$(arg scan_20_hz)" name="max_clouds" type="int" value="10" />
      <param name="fixed_frame_id" type="string" value="" />
      <param name="queue_size" type="int" value="$(arg queue_size)" />
    </node> -->

    <!-- <node pkg="nodelet" type="nodelet" name="pointcloud_to_depthimage" args="standalone rtabmap_util/pointcloud_to_depthimage" output="screen">
      <remap from="cloud" to="$(arg scan_topic_deskewed)"/>
      <remap from="camera_info" to="/$(arg vehicle)_$(arg ID)/camera/rgb/camera_info"/>
      
      <param name="approx" value="false"/>
      <param name="fill_holes_size" value="2"/>

      <remap from="image_raw" to="/$(arg vehicle)_$(arg ID)/camera/depth/points/image_raw"/>
      <remap from="image" to="/$(arg vehicle)_$(arg ID)/camera/depth/points/image"/>
    </node> -->

  </group>

  <!-- rviz -->
  <node name="rviz" pkg="rviz" type="rviz" args="-d $(find drone_planning_ros)/launch/rviz/drone_planning_rtabmap.rviz" />

</launch>